From fa2d082628bc69b779ecd9937e7af693db092c19 Mon Sep 17 00:00:00 2001
From: Guoqing Jiang <Guoqing.Jiang@windriver.com>
Date: Wed, 17 Jul 2013 16:08:19 +0800
Subject: [PATCH 23/61] ARM: keystone: override dma address conversion helpers

Commit 0baf55c0b97baec84e52be8ae2dc6a47a4ccee8c from arago tree
git://arago-project.org/git/projects/linux-keystone.git releases/03.00.00.11/master

This patch adds keystone specific overrides for helper functions that convert
back and forth between dma_addr_t and physical/virtual addresses.

Signed-off-by: Cyril Chemparathy <cyril@ti.com>
Signed-off-by: Guoqing Jiang <Guoqing.Jiang@windriver.com>
---
 arch/arm/mach-keystone/include/mach/memory.h |   11 +++++++++++
 arch/arm/mach-keystone/keystone.c            |    5 +++++
 2 files changed, 16 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-keystone/include/mach/memory.h b/arch/arm/mach-keystone/include/mach/memory.h
index a5f7a1a..872183f5 100644
--- a/arch/arm/mach-keystone/include/mach/memory.h
+++ b/arch/arm/mach-keystone/include/mach/memory.h
@@ -40,6 +40,17 @@ static inline phys_addr_t __virt_to_idmap(unsigned long x)
 
 #define virt_to_idmap(x)	__virt_to_idmap((unsigned long)(x))
 
+extern unsigned long __arch_dma_pfn_offset;
+
+#define __arch_pfn_to_dma(dev, pfn)				\
+	PFN_PHYS(pfn - __arch_dma_pfn_offset)
+#define __arch_dma_to_pfn(dev, addr)				\
+	(PFN_DOWN(addr) + __arch_dma_pfn_offset)
+#define __arch_dma_to_virt(dev, addr)				\
+	phys_to_virt(addr + PFN_PHYS(__arch_dma_pfn_offset))
+#define __arch_virt_to_dma(dev, addr)				\
+	(virt_to_phys(addr) - PFN_PHYS(__arch_dma_pfn_offset))
+
 #endif /* __ASSEMBLY__ */
 
 #endif /* CONFIG_ARM_LPAE */
diff --git a/arch/arm/mach-keystone/keystone.c b/arch/arm/mach-keystone/keystone.c
index 9f40a0d..a4f1f7e 100644
--- a/arch/arm/mach-keystone/keystone.c
+++ b/arch/arm/mach-keystone/keystone.c
@@ -77,6 +77,8 @@ static const char *keystone_match[] __initconst = {
 	NULL,
 };
 
+unsigned long __arch_dma_pfn_offset;
+
 static void __init keystone_init_meminfo(void)
 {
 	bool lpae = IS_ENABLED(CONFIG_ARM_LPAE);
@@ -110,6 +112,9 @@ static void __init keystone_init_meminfo(void)
 	pr_info("switching to high address space at 0x%llx\n", offset);
 	__pv_phys_offset = offset;
 	__pv_offset	= offset - PAGE_OFFSET;
+
+	__arch_dma_pfn_offset = PFN_DOWN(KEYSTONE_HIGH_PHYS_START -
+					 KEYSTONE_LOW_PHYS_START);
 }
 
 DT_MACHINE_START(KEYSTONE, "KeyStone")
-- 
1.7.5.4

